Validate and normalize file name when adding entries to ZIP files

xavierleroy opened the pull request:
- The ZIP specification says that forward slashes must be used instead of backslashes in file names.  Transform `\` to `/` under Windows and reject `\` in file names on other systems.

- Also check that the file name is not absolute.

Fixes: #40 .

alainfrisch commented on zip.ml:
```diff
@@ -588,8 +588,16 @@ let add_entry_header ofile comment level mtime filename =
   if String.length comment >= 0x10000 then
     raise(Error(ofile.of_filename, filename, "comment too long"));
+  let filename =
+    if not (String.contains filename '\\') then filename
```
This implements the constraint from the spec, but I wonder if it could break some existing uses of the library. I tried under Linux, and zip/unzip don't complain about backslashes in file names (i.e. they don't respect the spec). One possibility could be to keep backslashes in filenames when not running on Windows; it's not nice, and the spec about the file format is not supposed to depend on which OS we used to create the file.

If we keep the strict semantics, I think it should be documented in the module interface and Changelog (as a possibly breaking changes for non-Windows users).

Another point: what about `Sys.os_type = "Cygwin"`?  In that case, one might also expect backslashes in filenames manipulated by the application.

(FTR, on Windows, unzipping a file with entry containing backslashes behaves as follows:
  - Using Cygwin's `unzip`: backslashes are turned into underscores.
  - Using WSL `unzip` (on an NTFS drive): backslashes are turned into the Unicode Private Use Area U+F05C.
)

alainfrisch commented on zip.ml:
```diff
@@ -588,8 +588,16 @@ let add_entry_header ofile comment level mtime filename =
   if String.length filename >= 0x10000 then
     raise(Error(ofile.of_filename, filename, "filename too long"));
+  if not (Filename.is_relative filename) then
+    raise(Error(ofile.of_filename, filename, "file name must not be absolute"));
```
The spec says:
> The path stored MUST NOT contain a drive or device letter, or a leading slash.

To be completely strict (and with idea that the constraint is about the resulting .zip file itself), one should reject e.g. `C:/foo` on all systems, not just Windows. To be noted: under Linux, `zip` happily creates a zip file with an entry named `c:\foo`.

xavierleroy commented:
The ZIP "specification" is an application note written in the 1980's with MS-DOS in mind and implemented with many useful(?) deviations, as you've noticed by testing with Info-zip (the most popular Unix implementation).  So, I read the paragraph you quoted twice as a recommendation to use forward slashes as directory delimiters, and avoid absolute file names.

But I agree it's the extractor that should check for malicious file names (absolute, `../../etc/passwd`, etc) and do something about it, not the creator of the ZIP file.  Indeed, Info-zip's `zip` command gladly stores absolute path names, but `unzip` strips the leading `/`.  At first,  I thought that checking at creation time would catch honest mistakes early, but perhaps it's too restrictive, so I removed that check.

Same reasoning for backslashes in file names when the ZIP file is not created under Windows: I think it's a mistake, as a Windows extractor must reject these file names, but maybe not worth reporting at creation time.

So, I've simplified this PR to just normalize directory separators if we're running under Windows, and nothing more.  Will merge next.

alainfrisch commented:
Thanks @xavierleroy !
